dll/win32/msgina/gui.c

```diff
@@ -985,12 +988,43 @@ SecurityDialogProc(
                     EndDialog(hwndDlg, WLX_SAS_ACTION_LOCK_WKSTA);
                     return TRUE;
                 case IDC_SECURITY_LOGOFF:
-                    if (OnLogOff(hwndDlg, pgContext) == IDYES)
+                    if (ISKEYDOWN(VK_CONTROL))
+                    {
+                        if (ResourceMessageBox(pgContext,
+                                               hwndDlg,
+                                               MB_OKCANCEL | MB_ICONSTOP,
+                                               IDS_FORCE_LOGOFF_TITLE,
+                                               IDS_FORCE_LOGOFF) == IDOK)
+                        {
+                            EndDialog(hwndDlg, WLX_SAS_ACTION_FORCE_LOGOFF);
+                        }
+                    }
+                    else if (OnLogOff(hwndDlg, pgContext) == IDYES)
+                    {
                         EndDialog(hwndDlg, WLX_SAS_ACTION_LOGOFF);
+                    }
                     return TRUE;
                 case IDC_SECURITY_SHUTDOWN:
-                    if (OnShutDown(hwndDlg, pgContext) == IDOK)
+                    /* Emergency restart feature */
+                    if (ISKEYDOWN(VK_CONTROL))
+                    {
+                        if (ResourceMessageBox(pgContext,
+                                               hwndDlg,
+                                               MB_OKCANCEL | MB_ICONSTOP,
+                                               IDS_EMERGENCY_RESTART_TITLE,
+                                               IDS_EMERGENCY_RESTART) == IDOK)
+                        {
+                            BOOLEAN Old;
+                            ERR("Emergency restarting NT...\n");
+                            RtlAdjustPrivilege(SE_SHUTDOWN_PRIVILEGE, TRUE, FALSE, &Old);
+                            NtShutdownSystem(ShutdownReboot);
+                            RtlAdjustPrivilege(SE_SHUTDOWN_PRIVILEGE, Old, FALSE, &Old);
```

Comment:
Probably the last thing to do, there was a tweet claiming "Restarting this way will cause Windows to break into a debugger if one is attached" (see Jira comment), this needs to be added.

Reply:
Added a `DebugBreak()` before shutting down the system. To do this I had to wrap it around a `__try/__except`. Without them the system never shuts down when a kernel debugger is not attached. However, this is incompatible with GCC so I disabled it when using that compiler. Clang should support the `__try/__except` macros.

Reply:
That's why we uniformly use our `_SEH2_TRY` and co. macros thorough our whole codebase.

Reply:
> Probably the last thing to do, there was a tweet claiming "Restarting this way will cause Windows to break into a debugger if one is attached" (see Jira comment), this needs to be added.

My reply in the JIRA report:
>{quote}
Also this tweet says "Restarting this way will cause Windows to break into a debugger if one is attached": https://twitter.com/XenoPanther/status/1648360869678465027
{quote}
>
> Hmmmmyeaaaaaah....  Well observe that this is in Windows 11+. Also the windbg session doesn't show who emitted the debugbreak. Could have been triggered from anywhere tbh. (like e.g. because it was a hard-reboot and ofc. some module didn't unload "as expected", etc.) So I would personally avoid adding extra explicit debug breaks in the PR.

Reply:
I've checked the debug build of Win2k3 msgina.dll and it doesn't emit any DebugBreak after the NtShutdownSystem call.

Reply:
So, are we calling `DebugBreak()` when a debugger is attached or not? Thumbs up for yes, thumbs down for no.